### ISSUE TITLE
No user attached to submission

### ISSUE THREAD

**Issue by romainpoirier** [verbb/formie#1618]
### Describe the bug

I have received around thirty submissions on a form.
For a reason I am not aware of, two submissions were received without an associated user.
However, the form is not anonymous and requires users to be logged in. Furthermore, users can't delete their account.
I cannot find any means to locate the user in question, and the logs do not provide any additional information.
What could have caused this? Is there a hidden way to locate these users?

<img width="372" alt="Capture d’écran 2023-11-02 à 13 04 48" src="https://github.com/verbb/formie/assets/4634192/c01f2835-30b4-4cf3-9adf-9af8f73ebd12">

### Steps to reproduce

1. Unknown

### Form settings

- Multi-page form: No
- Submission Method: Page Reload
- Client-side Validation: Yes
- Custom Form Templates: No

### Craft CMS version

Craft Pro 4.4.17

### Plugin version

2.0.34.1

### Multi-site?

Yes

**Comment by engram-design** [comment]
I assume you have “Collect User” enabled in your form settings? There’s no value in that column in the “formie_forms” db table?

that’s indeed a bit odd. For reference this is how we assign it https://github.com/verbb/formie/blob/e266a1e672c776790829f4d16017fd5635524962/src/controllers/SubmissionsController.php#L1148

**Comment by romainpoirier** [comment]
Yes, `Collect User` is on.
And in the `settings` column of the `formie_forms` value of this form in the DB I can see: `"collectUser": true`.

**Comment by romainpoirier** [comment]
On top of that, I can say that I have large performance issues on this form on front-end.
It sometimes requires more than 10 seconds to load it. Maybe that would be an explanation of this missing data?

To contextualize, this form has just two fields:

- One _Users_ field;
- One _Agree_ field;

The _Users_ is loading as sources all possible users of this website (~3,400).

I can see in the Usage tab of the form:

- The Entry where the form is added (using the Form field);
- The list of the thirty submissions of this form (Matrix Block);

Why are they listed there, for what use, and could it impact performance at some point?

Prior to that, the same form was working fine on Craft 3 + Formie 1.

**Comment by engram-design** [comment]
So it's a great point - element fields will load all the elements from their source in order to allow them to be pickable. For a dropdown Users field, we have to query all your users to pick from. There's no lazy loading or pagination when it comes to available options, so that might contribute to the slowness.

The usage stats won't affect performance, that's just calculated at the form builder.

But that shouldn't affect recording the user against the submission, as that's done when submitting the form. I can't really figure out why the user wouldn't be recorded against the submission as you describe, unless the user was indeed not logged in, or something else regarding the user session was incorrect. What's more frustrating is that it's not consistent.

I'll keep investigating to try and reproduce.

**Comment by romainpoirier** [comment]
> For a dropdown Users field, we have to query all your users to pick from. 

In my case, I have around 3,400 users to load. Is there any way to optimize this request by default?
To be more specific, I need to load all the users who have responded to another form (the registration form). In this second form, in the "Users" field, I need to load the selected users who would be able to receive meeting requests.
Do you think it would be possible to add this kind of feature?

> I can't really figure out why the user wouldn't be recorded against the submission as you describe, unless the user was indeed not logged in, or something else regarding the user session was incorrect.

Yes, it's possible that the user's session ended while submitting. Is it possible that it could have saved the submission anonymously instead of not submitting it at all?

**Comment by engram-design** [comment]
I’d recommend looking to modifying the user element query that’s used to generate the options for the field. It could be something as simple as limiting to a number, or something more advanced like you describe. https://verbb.io/craft-plugins/formie/user-guides/how-to-modify-element-fields-queries-to-change-what-elements-can-be-chosen

and that’s a possibility of the submission being saved without a user attached to it. That setting won’t require a user to be logged in (that’s another setting in behaviors).

**Comment by romainpoirier** [comment]
> https://verbb.io/craft-plugins/formie/user-guides/how-to-modify-element-fields-queries-to-change-what-elements-can-be-chosen

The problem is that even if it could be possible to modify the user element, it would be done from the template side, not from the end user which has only access to the control panel. So that's a bit challenging for a dev-free setup like mine, where the end client should be able to setup the forms by itself. I'll try to find out a trick to add specific rules based on the form handle.

> and that’s a possibility of the submission being saved without a user attached to it. That setting won’t require a user to be logged in (that’s another setting in behaviors).

How to disable this to prevent anonymous submission?

**Comment by engram-design** [comment]
Sure, we don't have an element query builder that's client-friendly to configure what elements are used for the User field. I was mainly referring to your comment on "I need to load all the users who have responded to another form", which is potentially something custom you need to do _just_ for this form? I don't have any plans to incorporate such advanced logic to element fields, but if there are other ways we can provide "filters" for what elements we have, I'm all ears.

I did consider adding a table field as a setting to define key-value element query params, but that seems pretty advanced for end-user clients who wouldn't know what to add there.

These were the settings I was referring to:
<img width="517" alt="image" src="https://github.com/verbb/formie/assets/1221575/615dc45d-668e-4d4f-8c93-6b664301ff63">

**Comment by romainpoirier** [comment]
> "I need to load all the users who have responded to another form", which is potentially something custom you need to do just for this form


To be more specific in my context, I have forms that allow users to register as participants for events. On each of these forms, there's a checkbox that allows them to register for the one-on-one part of the event, where they can meet others. We also have another form that is dedicated to one-on-one registration, where registered users can select who they want to meet at the event. In this case, the second form is linked to the previous one, allowing users to select the participants from the previous form they want to meet.

On the one-on-one form, I have a layer on top of the form that displays only the users related to the current targeted event. In the background, when you select one user or another from this layer, it will automatically select the user on the hidden Formie form.

Upon further testing my code, it appears that the performance issues stem from the [Retcon plugin](https://github.com/mmikkel/Retcon-Craft), which applies transformations to my `craft.formie.renderForm()`. With a large number of users (around 3,400) and when targeting elements like `.fui-input`, `.fui-checkbox`, and `.fui-checkbox-label`, which are abundant, these transformations seems to significantly slow down page rendering.

> These were the settings I was referring to:

Oh, thank you! I can confirm that the _Required Logged-in User_ was'nt already checked on these forms.

**Comment by engram-design** [comment]
Ah, that's interesting. I thought 10 seconds was a _little_ slow for just fetching that amount of users (although not unheard of for dev environments).

**Comment by romainpoirier** [comment]
> I’d recommend looking to modifying the user element query that’s used to generate the options for the field. It could be something as simple as limiting to a number, or something more advanced like you describe. https://verbb.io/craft-plugins/formie/user-guides/how-to-modify-element-fields-queries-to-change-what-elements-can-be-chosen

Sorry to coming back later or this point, but I'm still trying to fix this performance issue.

This is what I tried:
```
{% set otherFormSubmissions = craft.formie.submissions().form(handle).registerToOtherForm(true) %}
{% set users = otherFormSubmissions.all() | map(item => item.user) %}
{% set field = form.attendees %}
{% do field.setElementsQuery(users) %}

{{ craft.formie.renderForm(form) }}
```

But I get this error:
`Impossible to invoke a method ("setElementsQuery") on a null variable.`

Looks like I'm failing to load and edit the query for the field `attendees` of my form.

Any idea how to fix this?

**Comment by engram-design** [comment]
The way to fetch a field from a form would be through the field layout, so it's similar to any other element:

```twig
{% set field form.getFieldByHandle('attendees') %}
```

**Comment by romainpoirier** [comment]
Thank you, this is working:
`{% set field = form.getFieldByHandle('attendees') %}`

But now I'm stuck at this step:
`{% do field.setElementsQuery(users) %}`

The problem is that the expected parameter (`users`) is an ElementQuery:
`Cannot assign array to property verbb\formie\fields\formfields\Users::$elementsQuery of type ?craft\elements\db\ElementQuery`

In my case, as you can see I can only send an Array, filtered from the other form submissions:
`{% set users = otherFormSubmissions.all() | map(item => item.user) %}`

I then tried to convert it to a User query like this:
```
{% set users = otherFormSubmissions.all() | map(item => item.user.id) %}
{% set users = craft.users.id(users) %}
```

But now get this error:
`Setting unknown property: craft\elements\db\UserQuery::0`

Adding a `.all()` on the users query doesn't fix this, neither adding `.limit(null)`.

**Comment by engram-design** [comment]
I was going to say, that supplying an ElementQuery is the way to go, with `craft.users.id(users)`, but there's indeed an issue with that.

Fixed for the next release. To get this early, run `composer require verbb/formie:"dev-craft-4 as 2.0.43"`.

Note that you can either create a new element query:

```twig
{% set users = craft.users.id([1]) %}
{% do field.setElementsQuery(users) %}
```

Or, you can extend the existing query:

```twig
{% set users = field.getElementsQuery().id([1]) %}
{% do field.setElementsQuery(users) %}
```

The benefit of the latter, is that for things like `orderBy` and `limit` are respected in the field settings. But you can otherwise start from scratch with your custom query as per the first option.

**Comment by engram-design** [comment]
Fixed in [2.0.44](https://github.com/verbb/formie/releases/tag/2.0.44)
